pkg/kritis/review/validating_transport.go

```diff
@@ -101,51 +102,80 @@ func (avt *AttestorValidatingTransport) validatePublicKeyId(pubKey v1beta1.Publi
+func (avt *AttestorValidatingTransport) parsePublicKeys() []cryptolib.PublicKey {
+	attestorKeys := avt.Attestor.Spec.PublicKeys
+	numKeys := len(attestorKeys)
+	publicKeys := []cryptolib.PublicKey{}
+
+	for i, attestorKey := range attestorKeys {
+		if err := avt.validatePublicKey(attestorKey); err != nil {
 			// warning level because single key failure is something tolerable
 			glog.Warningf("Error parsing key %d (%d keys total) for %q: %v", i, numKeys, avt.Attestor.Name, err)
 			continue
 		}
-		if _, ok := keys[pubKey.KeyId]; ok {
-			glog.Warningf("Overwriting key with same fingerprint %s for %q.", pubKey.KeyId, avt.Attestor.Name)
+		decodedKey, err := base64.StdEncoding.DecodeString(attestorKey.AsciiArmoredPgpPublicKey)
+		if err != nil {
+			glog.Infof("Cannot base64 decode public key: %v", err)
+			continue
 		}
-		keys[pubKey.KeyId] = pubKey
-	}
-	if len(keys) == 0 {
-		return nil, fmt.Errorf("unable to find any valid key for %q", avt.Attestor.Name)
+		publicKey := cryptolib.NewPublicKey(cryptolib.Pgp, decodedKey, attestorKey.KeyId)
+		publicKeys = append(publicKeys, publicKey)
 	}
+	return publicKeys
+}
 
-	out := []attestation.ValidatedAttestation{}
-	host, err := container.NewAtomicContainerSig(image, map[string]string{})
-	if err != nil {
-		glog.Error(err)
-		return nil, err
-	}
+func (avt *AttestorValidatingTransport) fetchAttestations(image string) ([]*cryptolib.Attestation, error) {
+	atts := []*cryptolib.Attestation{}
 	rawAtts, err := avt.Client.Attestations(image, &avt.Attestor)
 	if err != nil {
-		glog.Error(err)
-		return nil, err
+		return nil, fmt.Errorf("error fetching attestations for image %s: %v", image, err)
 	}
 
 	for _, rawAtt := range rawAtts {
 		if rawAtt.SignatureType != metadata.PgpSignatureType {
-			return nil, fmt.Errorf("Signature type %s is not supported for Attestation %v", rawAtt.SignatureType.String(), rawAtt)
+			glog.Warningf("Skipping attestation with unsupported signature type %s", rawAtt.SignatureType.String())
+			continue
 		}
 		decodedSig, err := base64.StdEncoding.DecodeString(rawAtt.Signature.Signature)
 		if err != nil {
 			glog.Warningf("Cannot base64 decode signature for attestation %v. Error: %v", rawAtt, err)
 			continue
 		}
-		keyId := rawAtt.Signature.PublicKeyId
-		if err = host.VerifySignature(keys[keyId], string(decodedSig)); err != nil {
-			glog.Warningf("Could not find or verify attestation for attestor %s: %s", keyId, err.Error())
+		// TODO(https://github.com/grafeas/kritis/issues/505): Remove this
+		// after Kritis migrates to cryptolib.Attestation.
+		att := &cryptolib.Attestation{
+			PublicKeyID:       rawAtt.Signature.PublicKeyId,
+			Signature:         decodedSig,
+			SerializedPayload: rawAtt.SerializedPayload,
+		}
+		atts = append(atts, att)
+	}
+	return atts, nil
+}
+
+func (avt *AttestorValidatingTransport) GetValidatedAttestations(image string) ([]attestation.ValidatedAttestation, error) {
+	publicKeys := avt.parsePublicKeys()
```

Comment:
Consider changing parsePublicKeys() to return (keys, error).  There can be many errors, and any aggregation will work.  For example, just send back the first or the last one encountered.  Or even just ids of offending keys.

If you do this, then parsing errors won't be entirely swallowed, and it will be easier for user to debug.

Reply:
ditto on returning parsing errors.

I am not sure if we should pass error when there is at least one valid key, but definitely aggregate errors when no key is valid.

Reply:
I modified this to also return the key ID's for keys that could not be parsed. The caller will error if none of the keys were valid, and its error message references the ID's of the invalid keys. The specific error for each invalid key should be accessible from the logs.